### PR TITLE
Support runner upgrade messages

### PR DESCRIPTION
**Pull request by takost**
- Added as parameter of `GetMessageAsync` runner version to get it later in the broker.
- Added new type of message `RunnerRefreshMessage`. Behaviour of update is the same, but Timeout value in new message will be number of milliseconds (int value). C# json format of the timespan looks like `d.hours:minutes:seconds.milliseconds": https://learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/6.0/timespan-serialization-format#new-behavior. In other languages it can differ, so I have chosen to use more generic way to pass this value. 
- 
Related issue(s):
- https://github.com/github/actions-broker/issues/53 
- 
Related PRs:
- [ ] [Broker PR](https://github.com/github/actions-broker/pull/206)
- [ ] [Actions PR](https://github.com/github/actions-dotnet/pull/13442)

### CODE REVIEW

File: src/Runner.Listener/Runner.cs [hunk 1]
```diff
@@ -466,6 +466,43 @@ private async Task<int> RunAsync(RunnerSettings settings, bool runOnce = false)
...
+                                if (autoUpdateInProgress == false)
+                                {
+                                    autoUpdateInProgress = true;
+                                    var runnerUpdateMessage = JsonUtility.FromString<RunnerRefreshMessage>(message.Body);
+                                    var agentRefreshMessage = new AgentRefreshMessage(runnerUpdateMessage.AgentId, runnerUpdateMessage.TargetVersion, TimeSpan.FromMilliseconds(runnerUpdateMessage.Timeout));
+#if DEBUG
+                                    // Can mock the update for testing
+                                    if (StringUtil.ConvertToBoolean(Environment.GetEnvironmentVariable("GITHUB_ACTIONS_RUNNER_IS_MOCK_UPDATE")))
+                                    {
+
```
Comment (lokesh755) [inline src/Runner.Listener/Runner.cs:480]: nit: extra line

File: src/Sdk/DTWebApi/WebApi/RunnerRefreshMessage.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        [DataMember]
+        public Int32 AgentId
+        {
+            get;
+            private set;
+        }
+
+        [DataMember]
+        public int Timeout
```
Comment (TingluoHuang) [inline src/Sdk/DTWebApi/WebApi/RunnerRefreshMessage.cs:36]: `TimeoutInSecond`

File: src/Sdk/DTWebApi/WebApi/RunnerRefreshMessage.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        {
+        }
+
+        public RunnerRefreshMessage(
+            Int32 agentId,
+            String targetVersion,
+            int? timeout = null)
+        {
+            this.AgentId = agentId;
+            this.Timeout = timeout ?? TimeSpan.FromMinutes(60).Milliseconds;
```
Comment (TingluoHuang) [inline src/Sdk/DTWebApi/WebApi/RunnerRefreshMessage.cs:24]: do we need milliseconds? or seconds is enough.

File: src/Runner.Listener/Runner.cs [hunk 4]
```diff
@@ -466,6 +466,43 @@ private async Task<int> RunAsync(RunnerSettings settings, bool runOnce = false)
                                     Trace.Info("Refresh message received, skip autoupdate since a previous autoupdate is already running.");
                                 }
                             }
+                            else if (string.Equals(message.MessageType, RunnerRefreshMessage.MessageType, StringComparison.OrdinalIgnoreCase))
+                            {
+                                if (autoUpdateInProgress == false)
+                                {
+                                    autoUpdateInProgress = true;
+                                    var runnerUpdateMessage = JsonUtility.FromString<RunnerRefreshMessage>(message.Body);
+                                    var agentRefreshMessage = new AgentRefreshMessage(runnerUpdateMessage.AgentId, runnerUpdateMessage.TargetVersion, TimeSpan.FromMilliseconds(runnerUpdateMessage.Timeout));
```
Comment (TingluoHuang) [inline src/Runner.Listener/Runner.cs:475]: can we merge this into the if the block on 433 so you don't have to duplicate all the code?

basically,
```C#
if (string.Equals(message.MessageType, AgentRefreshMessage.MessageType, StringComparison.OrdinalIgnoreCase) ||
    string.Equals(message.MessageType, RunnerRefreshMessage.MessageType, StringComparison.OrdinalIgnoreCase))
{
   AgentRefreshMessage runnerUpdateMessage = null;
   if (string.Equals(message.MessageType, AgentRefreshMessage.MessageType, StringComparison.OrdinalIgnoreCase))
    {
         runnerUpdateMessage = JsonUtility.FromString<AgentRefreshMessage>(message.Body);
    }
    else
    {
       var brokerRunnerUpdateMessage = JsonUtility.FromString<RunnerRefreshMessage>(message.Body); 
       runnerUpdateMessage = new AgentRefreshMessage(brokerRunnerUpdateMessage.AgentId, brokerRunnerUpdateMessage.TargetVersion, TimeSpan.FromMilliseconds(brokerRunnerUpdateMessage.Timeout));
    }

    // existing code.
}
```

**Review by TingluoHuang — CHANGES_REQUESTED** [review]
Might need some cleanup. 😄

**Review by TingluoHuang — APPROVED** [review]

**Review by vmjoseph — APPROVED** [review]

**Pull request merged**